File: codes_c/fibmem.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>



//Receives n(arg[1]) from terminal-main, iterates step by step to n-th term
//Operations counter present. Memo array is received, and space to count operations.
// if the calculation is made it deposits in into the index so it gets spotted next time around.

long long fibmem(int n, long long *memo, int *opC) {
    if(n <=2) {
        memo[n] = 1;
    }
    if (memo[n] != 0 || n == 0) {
        return memo[n];
    }
    // recursive portion
    memo[n] = fibmem(n-1, memo, opC) + fibmem(n -2, memo, opC);
    (*opC)++;
    return memo[n];
}


//calculates passes an int from terminal after program call to function. clock starts
// clock stops. prints each term from from 1 to n. stops clock. prints op number
int main(int argc, char *argv[]) {
    int n = atoi(argv[1]);
    long long *memo = calloc(n +1, sizeof(int));
    int opC = 0;
    clock_t start, end;
    start = clock();
    for (int i = 1; i <= n; i++) {
        printf("term %d: %lld\n", i, fibmem(i, memo, &opC));
    }
    end = clock();
    double total = (double)(end-start) / CLOCKS_PER_SEC;
    free(memo);
    printf("Time: %fs\n", total);
    printf("Total Operations: %d\n", opC);
    return 0;
}

File: codes_rs/fibmem.rs
use std::collections::HashMap;

//Unlike C, Rust uses a Hashmap, but to me the process isnt wildly different.
//Receives n(arg[1]) from terminal-main, iterates step by step to n-th term
//Operations counter present. Memo array is received, and space to count operations.
// if the calculation is made it deposits in into the index of memo so it gets spotted next time around.
fn fibmem(n: u64, memo: &mut HashMap<u64, u64>, op_c: &mut u64) -> u64 { 
    if n <= 2 {return 1;}
    if let Some(val) = memo.get(&n) {
        return *val;
    }
    let current = fibmem(n-1, memo, op_c) + fibmem(n-2, memo, op_c);
    *op_c +=1;
    memo.insert(n, current);
    current
    }


    /* functionally no different from the c versions. */
fn main() {
    let args: Vec<String> = std::env::args().collect();
    let n: u64 = args[1].parse().unwrap();
    let mut memo = HashMap::new();
    let mut op_c: u64 = 0;
    let start = std::time::Instant::now();
    for i in 1..=n {
        println!("term {}: {}", i, fibmem(i, &mut memo, &mut op_c));
    }
    let duration = start.elapsed();
    println!("Time: {}", duration.as_secs_f64());
    println!("Total Operations: {}", op_c);
}